### PR TITLE
feat: [PL-41271]: DelegateDown and DelegateExpiry Notification 

### PR DESCRIPTION
**Pull request by jennyjames21**
## Describe your changes

Summary:
-Add Notification to DelegateExpiry and DelegateDown events

Test
-Unit test
- End-to-end test to notification on delegate down event
- Verified delegateExpire set correctly in DelegateGroup collection
image

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/54519)

### CODE REVIEW

**Comment by jennyjames21** [comment]
trigger uts

**Comment by jennyjames21** [comment]
trigger uts

**Comment by jennyjames21** [comment]
trigger commonchecks

File: 400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java [hunk 1]
```diff
@@ -3263,6 +3264,12 @@ public DelegateGroup upsertDelegateGroup(String name, String accountId, Delegate
     if (sizeDetails != null) {
       setUnset(updateOperations, DelegateGroupKeys.sizeDetails, sizeDetails);
     }
+    if (delegateSetupDetails != null && isNotEmpty(delegateSetupDetails.getVersion())) {
+      long delegateExpiry = delegateSetupService.getDelegateExpirationTime(
```
Comment (xingchi-jin) [inline 400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java:3268]: it's probably better not to create a new var delegateExpirOn, we can calculate the expiry time on demand

**Comment by jennyjames21** [comment]
trigger uts

**Comment by jennyjames21** [comment]
*[`400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java` line 3268 at r2](https://reviewable.io/reviews/harness/harness-core/54519#gh-1385743728:-NilRlg16N9OsErIJLpp:b-9bpv9o) ([raw file](https://github.com/harness/harness-core/blob/c179e183b8a8112c508430c8f0ed1ae52be90e38/400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java#L3268)):*
<details><summary><i>Previously, xingchi-jin (Xingchi Jin) wrote…</i></summary><blockquote>

it's probably better not to create a new var delegateExpirOn, we can calculate the expiry time on demand
</blockquote></details>

as discussed over the call, we will keep it in the DB for iterator purpose

File: 400-rest/src/main/java/io/harness/iterator/DelegateExpiryAlertIterator.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+      PersistenceIteratorFactory.RedisBatchExecutorOptions executorOptions, Duration targetInterval) {
+    iterator = (MongoPersistenceIterator<DelegateGroup, MorphiaFilterExpander<DelegateGroup>>)
+                   persistenceIteratorFactory.createRedisBatchIteratorWithDedicatedThreadPool(executorOptions,
+                       DelegateGroup.class,
+                       MongoPersistenceIterator.<DelegateGroup, MorphiaFilterExpander<DelegateGroup>>builder()
+                           .clazz(DelegateGroup.class)
+                           .fieldName(DelegateGroupKeys.delegateExpiryAlertNextIteration)
+                           .filterExpander(q
+                               -> q.field(DelegateGroupKeys.delegatesExpireOn)
+                                      .lessThan(System.currentTimeMillis() - TimeUnit.DAYS.toMillis(45)))
```
Comment (xingchi-jin) [inline 400-rest/src/main/java/io/harness/iterator/DelegateExpiryAlertIterator.java:72]: what does this filter mean ? why 45 days

File: 400-rest/src/main/java/io/harness/iterator/DelegateExpiryAlertIterator.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+  @Override
+  public void registerIterator(IteratorExecutionHandler iteratorExecutionHandler) {
+    iteratorName = "DelegateExpiryAlert";
+    iteratorExecutionHandler.registerIteratorHandler(iteratorName, this);
+  }
+
+  @Override
+  public void handle(DelegateGroup delegateGroup) {
+    if (isDelegateExpiryNotificationAlreadySent(delegateGroup)) {
```
Comment (xingchi-jin) [inline 400-rest/src/main/java/io/harness/iterator/DelegateExpiryAlertIterator.java:87]: this implementation depends on the "SLA" is notification. 
do we require every expiry event to deliver exactly once, or at least once ?

File: 400-rest/src/main/java/io/harness/iterator/DelegateDisconnectDetectorIterator.java [hunk 4]
```diff
@@ -140,11 +151,39 @@ public void handle(Delegate delegate) {
       delegateDao.delegateDisconnected(delegate.getAccountId(), delegate.getUuid());
       delegateService.updateLastExpiredEventHeartbeatTime(
           delegate.getLastHeartBeat(), delegate.getUuid(), delegate.getAccountId());
+      if (featureFlagService.isEnabled(DEL_NOTIFICATION, delegate.getAccountId())) {
+        triggerNotificationRequest(delegate);
```
Comment (xingchi-jin) [inline 400-rest/src/main/java/io/harness/iterator/DelegateDisconnectDetectorIterator.java:155]: what if delegate unregister, and delete delegate, do we send notification "DOWN" ?

File: 400-rest/src/main/java/io/harness/iterator/DelegateExpiryAlertIterator.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    templateData.put("TEMPLATE_IDENTIFIER", "delegate_expired");
+    templateData.put("DELEGATE_NAME", delegateGroup.getIdentifier());
+    NotificationTriggerRequest.Builder notificationTriggerRequestBuilder =
+        NotificationTriggerRequest.newBuilder()
+            .setId(notificationTriggerRequestId)
+            .setAccountId(delegateGroup.getAccountId())
+            .setOrgId(orgId)
+            .setProjectId(projectId)
+            .setEventEntity(NotificationEntity.DELEGATE.name())
+            .setEvent(NotificationEvent.DELEGATE_DOWN.name())
```
Comment (raghuAtWings) [inline 400-rest/src/main/java/io/harness/iterator/DelegateExpiryAlertIterator.java:109]: why is this delegate down?

File: 400-rest/src/main/java/io/harness/iterator/DelegateExpiryAlertIterator.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+      PersistenceIteratorFactory.PumpExecutorOptions executorOptions, Duration targetInterval) {
+    iterator =
+        (MongoPersistenceIterator<DelegateGroup, MorphiaFilterExpander<DelegateGroup>>)
+            persistenceIteratorFactory.createPumpIteratorWithDedicatedThreadPool(executorOptions, DelegateGroup.class,
+                MongoPersistenceIterator.<DelegateGroup, MorphiaFilterExpander<DelegateGroup>>builder()
+                    .clazz(DelegateGroup.class)
+                    .fieldName(DelegateGroupKeys.delegateExpiryAlertNextIteration)
+                    .filterExpander(q
+                        -> q.field(DelegateGroupKeys.delegatesExpireOn)
+                               .lessThan(System.currentTimeMillis() - TimeUnit.DAYS.toMillis(45)))
```
Comment (raghuAtWings) [inline 400-rest/src/main/java/io/harness/iterator/DelegateExpiryAlertIterator.java:52]: don't understand this 45 days logic. shouldn't we keep sending update every day the delegate is expired?

File: 400-rest/src/main/java/io/harness/iterator/DelegateExpiryAlertIterator.java (hunk 3 above)
Comment (raghuAtWings) [reply 400-rest/src/main/java/io/harness/iterator/DelegateExpiryAlertIterator.java:87]: should we make this simpler by sending one notification every day. What if the notification is missed and then we never notify. i think we should run this iterator every day and notify for every expired delegate.

**Review by raghuAtWings — CHANGES_REQUESTED** [review]

File: 400-rest/src/main/java/io/harness/iterator/DelegateDisconnectDetectorIterator.java (hunk 4 above)
Comment (jennyjames21) [reply 400-rest/src/main/java/io/harness/iterator/DelegateDisconnectDetectorIterator.java:155]: I'm not sure that's the same use case here. I can check with pratheek
1) Rolling upgrade or version upgrade or autoscaling
 2) delete on purpose

Another thing we use delegate to send notifications to email/slack if delegate is down we are relying on other delegates in the account 😆

File: 400-rest/src/main/java/io/harness/iterator/DelegateExpiryAlertIterator.java (hunk 6 above)
Comment (jennyjames21) [reply 400-rest/src/main/java/io/harness/iterator/DelegateExpiryAlertIterator.java:52]: im using this iterator for both delegateExpiry and delegateAboutToExpire. 
DelegateaboutToExpire event allows to send notifications 1-4 weeks before expiration.

Comment (jennyjames21) [reply 400-rest/src/main/java/io/harness/iterator/DelegateExpiryAlertIterator.java:52]: i will change to 30 days

File: 400-rest/src/main/java/io/harness/iterator/DelegateExpiryAlertIterator.java (hunk 2 above)
Comment (jennyjames21) [reply 400-rest/src/main/java/io/harness/iterator/DelegateExpiryAlertIterator.java:72]: im using this iterator for both delegateExpiry and delegateAboutToExpire.
DelegateaboutToExpire event allows to send notifications 1-4 weeks before expiration.

File: 400-rest/src/main/java/io/harness/iterator/DelegateExpiryAlertIterator.java (hunk 3 above)
Comment (jennyjames21) [reply 400-rest/src/main/java/io/harness/iterator/DelegateExpiryAlertIterator.java:87]: interval to run in iterator in iterator_config.json
 "targetIntervalInSeconds": 86400,

**Comment by jennyjames21** [comment]
trigger uts

File: 400-rest/src/main/java/io/harness/iterator/DelegateExpiryAlertIterator.java (hunk 6 above)
Comment (xingchi-jin) [reply 400-rest/src/main/java/io/harness/iterator/DelegateExpiryAlertIterator.java:52]: should this be System.currentTimeMillis() + TimeUnit.DAYS.toMillis(30)

Comment (xingchi-jin) [reply 400-rest/src/main/java/io/harness/iterator/DelegateExpiryAlertIterator.java:52]: in addition, is it possible to make this number "30" a configuration, i m afraid we will run into customer requests where people want to change
or should we do per account config

Comment (jennyjames21) [reply 400-rest/src/main/java/io/harness/iterator/DelegateExpiryAlertIterator.java:52]: yes it should be +

It's configurable already, we just fetching the record from DB and send the notification only based on their settings
That check will be on the next iteration due to dependency to another PR

File: 400-rest/src/main/java/io/harness/iterator/DelegateExpiryAlertIterator.java (hunk 5 above)
Comment (jennyjames21) [reply 400-rest/src/main/java/io/harness/iterator/DelegateExpiryAlertIterator.java]: for testing, updated

**Comment by jennyjames21** [comment]
trigger uts

**Comment by jennyjames21** [comment]
trigger commonchecks

**Comment by jennyjames21** [comment]
trigger commonchecks

**Comment by jennyjames21** [comment]
trigger uts

**Comment by jennyjames21** [comment]
trigger uts

**Comment by jennyjames21** [comment]
trigger uts

**Review by raghuAtWings — APPROVED** [review]

**Pull request merged**